packages/functional-tests/lib/metrics.ts

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Page } from '@playwright/test';
+
+const GlobalEventProperties = [
+  'country_code',
+  'event_type',
+  'language',
+  'country_code_source',
+  'time',
+  'user_id',
+];
+
+const SubscriptionCreateEventProperties = [
+  'entrypoint',
+  'entrypoint_experiment',
+  'entrypoint_variation',
+  'form_type',
+  'utm_campaign',
+  'utm_content',
+  'utm_term',
+  'utm_medium',
+  'context',
+  'checkout_type',
+  'error_id',
+  'flow_id',
+  'payment_provider',
+  'plan_id',
+  'product_id',
+  'promotion_code',
+  'subscribed_plan_ids',
+];
+
+const SubscriptionUpgradeEventProperties = [
+  'previous_plan_id',
+  'previous_product_id',
+  'subscription_id',
+];
+
+const SubscriptionCancelledEventProperties = [
+  'payment_provider',
+  'plan_id',
+  'product_id',
+  'provider_event_id',
+  'subscription_id',
+  'voluntary_cancellation',
+];
+
+interface MetricsEventsDictionary<T> {
+  [key: string]: string;
+}
+
+export class MetricsObserver {
+  constructor(
+    public page: Page,
+    private eventsDictionary: MetricsEventsDictionary<string | number>,
+    private events: string[]
+  ) {}
+
+  // post request to /metrics to track all events
+  trackMetrics() {
+    this.page.on('request', (request) => {
+      if (request.method() === 'POST' && request.url().includes('/metrics')) {
+        const data = request.postDataJSON();
+
+        for (const [key, value] of Object.entries(data.events)) {
+          if (GlobalEventProperties.includes(key)) {
+            if (key !== 'event_type') {
+              eventsDictionary[key] = value;
+              events.push(value);
+            }
+
+            // TODO: check event type and add additional properties
+          }
+        }
+      }
+    });
+  }
+
+  // return metrics array
+  getMetrics() {
+    if (events.length > 0) {
+      return events;
+    }
+
+    return;
+  }
+
+  // verify events appear in correct order
+  verifyEventsOrder() {
+    // TODO: check order
+  }
+
+  // check if particular event exists
+  verifyEventExists(event: number | string) {
+    return event in eventsDictionary;
+  }
+
+  // get event error if it exists
+  getEventError() {
+    if (eventsDictionary.error_id) {
+      return eventsDictionary.error_id;
+    }
+
+    return;
+  }
```

Comment:
Hmm I am not sure I follow why we need some of these methods, like `getEventError` and `verifyEventExists`. What did you have in mind?